core/utils/STree.py

```diff
@@ -55,20 +55,64 @@ def compute_s_link(s: str, u: Node) -> None:
     u.s_link = v
 
 
+def traverse_edge(text: str, pattern: str, idx: int, start: int, end: int) -> tuple[int, int]:
+    i: int = start
+
+    while i < end and idx < len(pattern):
+        if text[i] != pattern[idx]:
+            return -1, idx
+
+        i += 1
+        idx += 1
+
+    # we have a match
+    if idx == len(pattern):
+        return 1, idx
+
+    # 0 means there are still more characters to match
+    return 0, idx
+
+
+def perform_search(node: Node, text: str, pattern: str, idx: int) -> int:
+    # case where the pattern starts with same alphabet as text
+    if node.idx == 0 and node.depth >= len(pattern):
+        return text.startswith(pattern)
+
+    if node.idx != 0:
+
+        """
+        We will traverse the edge and match alphabet by alphabet
+        The edge label has length = end - start + 1, where
+        start = node.idx + node.parent.depth
+        end = node.idx + node.depth
+        """
+        search_res, idx = traverse_edge(text, pattern, idx, node.idx + node.parent.depth, node.idx + node.depth)
+
+        if search_res != 0:
+            return search_res
+
+    if pattern[idx] in node.children:
+        return perform_search(node.children[pattern[idx]], text, pattern, idx)
```

Comment:
This is a tail-call, but in Python that is still a function call and you will run out of stack space if you are not careful. It is trivial to make into a loop, though.

Reply:
I will refactor this in the next PR. Thanks for the heads up 👍🏼